pkg/controller/backup_storage_location_controller.go

```diff
@@ -0,0 +1,210 @@
+}
+
+func NewBackupStorageLocationController(
+	namespace string,
+	defaultBackupLocation string,
+	defaultStoreValidationPeriod time.Duration,
+	backupLocationClient velerov1client.BackupStorageLocationsGetter,
+	backupStorageLocationLister velerov1listers.BackupStorageLocationLister,
+	newPluginManager func(logrus.FieldLogger) clientmgmt.Manager,
+	logger logrus.FieldLogger,
+) Interface {
+	if defaultStoreValidationPeriod <= 0 {
+		defaultStoreValidationPeriod = time.Minute
+	}
+
+	logger.Infof("Backup location store validation period is %v", defaultStoreValidationPeriod)
+
+	c := backupStorageLocationController{
+		genericController:            newGenericController("backup-storage-location", logger),
+		namespace:                    namespace,
+		defaultBackupLocation:        defaultBackupLocation,
+		defaultStoreValidationPeriod: defaultStoreValidationPeriod,
+		backupLocationClient:         backupLocationClient,
+		backupStorageLocationLister:  backupStorageLocationLister,
+
+		// use variables to refer to these functions so they can be
+		// replaced with fakes for testing.
+		newPluginManager: newPluginManager,
+		newBackupStore:   persistence.NewObjectBackupStore,
+	}
+
+	c.resyncFunc = c.run
+	c.resyncPeriod = 30 * time.Second
+
+	return c
+}
+
+func (c *backupStorageLocationController) run() {
+	c.logger.Info("Checking for existing locations ready to be validated; there needs to be at least 1 location that is available")
+
+	locations, err := c.backupStorageLocationLister.BackupStorageLocations(c.namespace).List(labels.Everything())
+	if err != nil {
+		c.logger.WithError(err).Error("Error listing locations, at least one available location is required")
+		return
+	}
+
+	if len(locations) == 0 {
+		c.logger.Error("No locations found, at least one available location is required")
+		return
+	}
+
+	pluginManager := c.newPluginManager(c.logger)
+	defer pluginManager.CleanupClients()
+
+	var unavailable []string
+	log := c.logger
+	for _, location := range locations {
+		locationName := location.Name
+		log = c.logger.WithField("backupLocation", locationName)
+
+		storeValidationPeriod := c.defaultStoreValidationPeriod
+		if location.Spec.StoreValidationPeriod != nil {
+			storeValidationPeriod = location.Spec.StoreValidationPeriod.Duration
+			if storeValidationPeriod == 0 {
+				log.Debug("Validation period for this location is set to 0, skipping validation")
+				continue
+			}
+
+			if storeValidationPeriod < 0 {
+				log.Debug("Validation period must be non-negative")
+				storeValidationPeriod = c.defaultStoreValidationPeriod
+			}
+		}
+
+		lastValidation := location.Status.LastValidationTime
+		if lastValidation != nil {
+			log.Debug("Checking if location needs to be validated at this time")
+			nextValidation := lastValidation.Add(storeValidationPeriod)
+			if time.Now().UTC().Before(nextValidation) {
+				continue
+			}
+		}
+
+		log.Debug("Location is ready to be validated")
```

Comment:
```suggestion
		log.Debug("Validating BackupStorageLocation")
```

Reply:
suggest using `BSL` instead of `location`.

Reply:
Ok I'll change these to `backup location` as it's used in other controllers. I'm trying to keep from using yet another way of referring to this resource!